### ISSUE TITLE
memory leak in MultiIndex

### ISSUE THREAD

**Issue by vograno** [pandas-dev/pandas#23035]
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import pandas as pd
import numpy as np
import gc
import psutil

def totmem(p):
    # total memory used by process in MB
    info = p.memory_info()
    return 1e-6*(info.vms + info.rss)

dat_size = 6000000
# make dataset with no data
# uncomment for regular index
# dat = pd.DataFrame(index=np.arange(dat_size))
# uncomment for MultiIndex index
dat = pd.DataFrame(index=pd.MultiIndex.from_arrays((np.arange(dat_size), np.arange(dat_size))))
# make bool vector for subsetting
sub = np.ones(dat_size, dtype=bool)
# init psutil stuff
p = psutil.Process()
gc.collect()
ram = totmem(p)

for i in range(10):
    dat.iloc[sub, :]     # leak happens here
    gc.collect()
    print(int(totmem(p) - ram))
    ram = totmem(p)

```
#### Problem description

this program leaks at a rate of 191 MB / cycle; it eventually runs out of memory if the loop goes on indefinitely.
In the program we use a bool vector to subset a dataset. If the dataset index is MultiIndex we observe a memory leak (as reported by the print statement). If the index is a regular mono-index no such leak is observed.

#### Expected Output
* the program should not run out memory.
* every output line should be zero.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
python: 3.6.6.final.0
python-bits: 64
OS: Linux
pandas: 0.23.4
numpy: 1.15.1

</details>

**Comment by dhimmel** [comment]
@vograno looks like the same memory leak issue I reported in https://github.com/pandas-dev/pandas/issues/23047? You posted this issue one day before mine, but I didn't see it until now!

> In the program we use a bool vector to subset a dataset. If the dataset index is MultiIndex we observe a memory leak (as reported by the print statement). If the index is a regular mono-index no such leak is observed.

This is consistent with our `tracemalloc` analysis that pinpointed the memory allocation to the following [lines](https://github.com/pandas-dev/pandas/blob/v0.23.4/pandas/core/indexes/multi.py#L2681-L2691) in the `MultiIndex.equals` function:

```python
# pandas/core/indexes/multi.py:2683
slabels = slabels[slabels != -1]
# pandas/core/indexes/multi.py:2688
olabels = olabels[olabels != -1]
```

Given that we reported the issue a day apart, I wonder if this issue is a regression and has not always been in Pandas? Tagging @gfyoung...

@vograno any workarounds you've found would also be appreciated!

**Comment by jeybrahms** [comment]
The only workaround I could think of is to use reset_index before and after subsetting.
Another option is to ditch multi-index altogether: keep the most important level in a mono-index and move the rest into the dataframe itself.
I'm waiting for pandas-dev to assign a milestone to see if I can wait the issue out.

**Comment by vograno** [comment]
@dhimmel The only workaround I could think of is to use reset_index before and after subsetting.
Another option is to ditch multi-index altogether: keep the most important level in a mono-index and move the rest into the dataframe itself.
I'm waiting for pandas-dev to assign a milestone to see if I can wait the issue out.

**Comment by dhimmel** [comment]
# Introduced in 0.24.4

Looks to me like the leak was introduced in Pandas 0.24.4. When I run @vograno's example code above (for 20 iterations) in pandas 0.23.3 + python 3.6.6, I get:

```
204, 48, 144, -108, 108, -108, 108, -108, 108, -108, 107, -107, 107, -107, 107, -107, 107, -107, 107, -107, -107
```

Each line here shows the difference in memory allocation in MB from the previous iteration. In other words, allocation stabilizes after a few iterations. However, with pandas 0.23.4 + python 3.6.6, I get the following:

```
336, 239, 336, 192, 192, 192, 191, 192, 192, 192, 191, 191, 192, 192, 191, 191, 192, 192, 192, 191, 
```

@jorisvandenbossche I see you contributed the last two changes to `pandas/core/indexes/multi.py` ([history](https://github.com/pandas-dev/pandas/commits/v0.23.4/pandas/core/indexes/multi.py)). However, these commits seem to have been part of Pandas since 0.23.2, so possibly the change was elsewhere. Any help appreciated.

**Comment by jreback** [comment]
@vograno well this would take a community contribution
no milestones are assigned until then have an actual mergable pull request

**Comment by vograno** [comment]
@dhimmel a downgrade to 0.23.3 (using pip) did NOT fix the problem for me, the program output remains the same. Could it be a C library that pip didn't downgrade (I treat pip as black magic)?
INSTALLED VERSIONS
------------------
python: 3.6.6.final.0
pandas: 0.23.3
numpy: 1.15.1

**Comment by dhimmel** [comment]
> a downgrade to 0.23.3 (using pip) did NOT fix the problem for me

Interesting. Not sure if a pip downgrade would also downgrade the underlying C libs, but I'd think so. 

This makes me think that the problem may be based on the versions of one of pandas dependencies... perhaps numpy. The two environments I compared were not well controlled: they had other differences than just pandas. I will play around with some more controlled conda environments and try to pinpoint when this issue was introduced.

**Comment by dhimmel** [comment]
Okay, I created the following four conda environments (my configuration defaults to using the conda-forge channel). All four environments use pandas 0.23.4 and python 3.6.6 but vary the numpy version.

```
conda create --yes --name=pandas-gh23035-env1 python=3.6.6 numpy=1.14.6 pandas=0.23.4 psutil
conda create --yes --name=pandas-gh23035-env2 python=3.6.6 numpy=1.15.0 pandas=0.23.4 psutil
conda create --yes --name=pandas-gh23035-env3 python=3.6.6 numpy=1.15.1 pandas=0.23.4 psutil
conda create --yes --name=pandas-gh23035-env4 python=3.6.6 numpy=1.15.2 pandas=0.23.4 psutil
```

I ran @vograno's script and evaluated whether there was a leak. The results are:

`pandas-gh23035-env1`: no leak
`pandas-gh23035-env2`: LEAK
`pandas-gh23035-env3`: LEAK
`pandas-gh23035-env4`: LEAK

Hence, I think it's very likely that the leak was introduced in numpy 1.15.0, which was a [hefty release](https://docs.scipy.org/doc/numpy-1.15.0/release.html). I was hoping the leak would subside in numpy [1.15.2](https://github.com/numpy/numpy/blob/04c76c58d6fcbb2031cc3392fcf88c8dce5f33c9/doc/release/1.15.2-notes.rst), which did fix a memory leak issue: https://github.com/numpy/numpy/issues/11951... although seemingly a different one than what's causing the pandas MultiIndex subset leak.

**Comment by vograno** [comment]
I can confirm a downgrade to numpy=1.14.6 fixes the problem. Thank you very much @dhimmel !

**Comment by dhimmel** [comment]
In https://github.com/numpy/numpy/issues/12122#issuecomment-430293579, @charris commented that this leak could be https://github.com/numpy/numpy/issues/12037 titled "Memory leak when slicing derived classes from Numpy", which was fixed in https://github.com/numpy/numpy/pull/12127 and will be in the numpy 0.15.3 release.

Seems like this could be the issue here as we've traced the issue to slicing multiindexes. If we do confirm this to be the issue, would we want to implement a RuntimeWarning in Pandas if the affected numpy version is in use and a MultiIndex is in use?
